### ISSUE TITLE
One case of #4973 not resolved

### ISSUE THREAD

**Issue by bebbi** [serverless/serverless#5011]
After more testing of https://github.com/serverless/serverless/pull/4989 (for #4973), looks like the fallback case might not be resolved.

The repro is simple. @erikerikson again sorry for delay.

serverless.yml:
```
provider:
  stage: ${opt:stage, self:custom.defaultStage}
  region: ${self:custom.region.region}
custom:
  defaultStage: staging
  region: ${file(./serverless-region.yml)}
```

serverless-region.yml:
```
regions:
  dev: us-east-1
# default eu-central-1
region: ${self:custom.region.regions.${self:provider.stage}, 'eu-central-1'}
```
This works for `serverless print --stage dev` but not for `serverless print` where the fallback is used.

`serverless print` output is:
`A valid undefined to satisfy the declaration 'deep:0.regions.staging' could not be found.`

Same behavior when the fallback value is a string or a reference:
```
regions:
  dev: us-east-1
  other: eu-central-1
region: ${self:custom.region.regions.${self:provider.stage}, self:custom.region.regions.other}
```

**Comment by bebbi** [comment]
Here's another case that looks a bit shady in `serverless print`:

serverless.yml: like before, and:
```
  custom:
    secretsConf: ${file(./serverless-secretsConf.yml)}
    serverlessSecrets: ${self:custom.secretsConf.serverlessSecrets}
```

with serverless-secretsConf.yml:
```
serverlessSecrets:
  providerOptions:
    region: ${self:provider.region}
```

resulting in serverless print output of:
```
  secretsConf:
    serverlessSecrets:
      [... correct structure here...]
    serverlessSecrets:
      $ref: '$["custom"]["secretsConf"]["serverlessSecrets"]'
```
Shouldn't this expand here?

**Comment by erikerikson** [comment]
Thank you very much for the reproduction cases @bebbi!  They are very helpful and I will add test cases and solve for them.

Regarding your second comment (containing `$ref`), that's actually the expected output.  That strange reference is output of json-cycle (https://github.com/serverless/serverless/blob/master/lib/plugins/print/print.js#L95) and how it deals with the circular reference you've created by having your service refer to an object in itself.  In the end, that section of the service is not written to the CloudFormation template so the circularity is not problematic and occasionally quite helpful (e.g. for referential consistency).

**Comment by bebbi** [comment]
Thanks @erikerikson.
After working around the above fallback issue, `serverless print` now works and the output looks reasonable, but `serverless deploy` still throws errors like:

`Trying to populate non string value into a string for variable ${deep:3}`

What's a simple way to troubleshoot this?

I've blindly tried to narrow it down, so e.g. when I hard-coded the region:
```
provider:
  region: 'eu-central-1'
```

the error is now
`Trying to populate non string value into a string for variable ${deep:0}`
instead of `deep:3`.

I'm not sure anymore that this issue is only about the fallback mechanism, but I'd appreciate some hints for quick troubleshooting!

**Comment by bebbi** [comment]
Update - it all ended up related to fallbacks.

**Comment by erikerikson** [comment]
Using the following:

serverless.yml
```
service: foo
provider:
  stage: ${opt:stage, self:custom.defaultStage}
  region: ${self:custom.region.region}
custom:
  defaultStage: staging
  region: ${file(./serverless-region.yml)}
```

serverless-region.yml
```
regions:
  dev: us-east-1
  other: us-other
# default eu-central-1
region: ${self:custom.region.regions.${self:provider.stage}, 'eu-central-1'}
region2: ${self:custom.region.regions.${self:provider.stage}, self:custom.region.regions.other}
```

here's the output I see:
```
$ sls print
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:0.regions.staging' could not be found.
 
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:0.regions.staging' could not be found.
 
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:0.regions.staging' could not be found.
 
service: foo
provider:
  stage: staging
  region: eu-central-1
custom:
  defaultStage: staging
  region:
    regions:
      dev: us-east-1
      other: us-other
    region: eu-central-1
    region2: us-other
```

The messages seem misleading and wrong though the result seems correct regardless.  Is this consistent with your observations?  Am I reproducing it properly?

This message seems correct, if hard to understand, in that `${self:custom.region.regions.${self:provider.stage}` resolves to `${self:custom.region.regions.staging}` since `${opt:stage, self:custom.defaultStage}` resolves to `staging` and `custom.region.regions.staging` is not defined.  The message, correct though it may be, could fairly be considered misleading because in the context of the default/fallback it is expected that the variable may not be resolvable.

In light of this, it appears that the bug here is that the warning message is produced despite being in the context of a default-able variable statement.  Do I understand this properly?

Thanks!

**Comment by bebbi** [comment]
@erikerikson No, while I understand your case, this is not the behavior I see. If I remove my `.staging` key in regions and rely on a fallback value, I get an error and no output at all.

I have a -very slightly- changed config to above, not sure that changes anything (plus a plural).

```
# serverless.yml
provider:
  region: ${self:custom.region}
custom:
  regions: ${file(../../serverless-regions.yml)}
  # Doing this to be explicit about the "imports" actually used in this file. The regions is used across services.
  region: ${self:custom.regions.region}
```

and

```
# serverless-regions.yml

regions:
  dev: us-east-1
  staging: eu-central-1
  test: eu-central-1
  production: eu-central-1
  other: eu-central-1

region: ${self:custom.regions.regions.${self:provider.stage}, self:custom.regions.regions.other}
```

The error is
```
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:1.regions.staging' could not be found.
 
[Repeated a few times] 

  Serverless Error ---------------------------------------
 
  Trying to populate non string value into a string for variable ${deep:1.regions.staging}. Please make sure the value of the property is a string.
```

**Comment by erikerikson** [comment]
Are you using `--stage staging` in the command?

**Comment by bebbi** [comment]
Yes. I had gotten rid of the fallback stage :D

**Comment by erikerikson** [comment]
Thank you, that may well explain the lack of repro.

**Comment by erikerikson** [comment]
Hi @bebbi.

TL;DR: I didn't get a reproduction of your issue. Please create a repository that reproduces the issue or close the issue please.

Here are the actions I took...

I copy-pasted your given text into their given named files.

I had to change:
```
regions: ${file(../../serverless-regions.yml)}
```
to
```
regions: ${file(serverless-regions.yml)}
```
and because I put the files in the same directory and then I had to add the following to your `serverless.yml` to pass validation:
```
service: foo
```
I then observe:
```
$ slsDev print --stage staging # the slsDev alias points to the ~/bin/serverless of my now out of date master

service: foo
provider:
  region: us-east-1
custom:
  regions:
    regions:
      dev: us-east-1
      test: eu-central-1
      production: eu-central-1
      other: eu-central-1
    region: us-east-1
  region: us-east-1
```

Given the expectation of an error, I tried this on the latest published version of serverless:
```
$ sls --version
1.27.3
```
And I see the same.  It seemed a little strange that us-east-1 was being used so I commented out the line containing `dev: us-east-1` in `serverless-regions.yml` to avoid the default stage value of `dev`.  I then observed:
```
$ sls print --stage staging
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:1.regions.dev' could not be found.
 
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:1.regions.dev' could not be found.
 
 
 Serverless Warning --------------------------------------
 
  A valid undefined to satisfy the declaration 'deep:1.regions.dev' could not be found.
 
service: foo
provider:
  region: eu-central-1
custom:
  regions:
    regions:
      test: eu-central-1
      production: eu-central-1
      other: eu-central-1
    region: eu-central-1
  region: eu-central-1
```

Here are my final copies of the files in their entireties:
```
# serverless.yml
service: foo
provider:
  region: ${self:custom.region}
custom:
  regions: ${file(serverless-regions.yml)}
  # Doing this to be explicit about the "imports" actually used in this file. The regions is used across services.
  region: ${self:custom.regions.region}
```

```
# serverless-regions.yml

regions:
  # dev: us-east-1
  # staging: eu-central-1
  test: eu-central-1
  production: eu-central-1
  other: eu-central-1

region: ${self:custom.regions.regions.${self:provider.stage}, self:custom.regions.regions.other}
```

I was similarly able to package and deploy.

Given a lack of reproduction using your materials, please provide a repository that reproduces the issue reliably if you'd like me to investigate further.

**Comment by erikerikson** [comment]
Perhaps additionally provide the information requested in the issue template regarding versions of node and serverless, et cetera.  Thanks!

**Comment by hojatbay** [comment]
Has this issue been fixed? I got the same problem after updating from 1.24.1 to 1.28.0.

**Comment by erikerikson** [comment]
@hojatbay can you please provide a sharable reproduction of the issue?  Thank you!

**Comment by hojatbay** [comment]
@erikerikson I created a sample sls project similar to my real project. If you try with 1.24.1 version you would be able to deploy this using "sls deploy --stage dev" command with no issue. But with version 1.27.1 and later you will see this error: "Profile ${deep:0.awsProfile} does not exist".

[my-service.zip](https://github.com/serverless/serverless/files/2220794/my-service.zip)

**Comment by erikerikson** [comment]
Hi @hojatbay - thanks for sharing materials that reproduce your issue.

If I change your
`variableSyntax: '\$\{((((self|opt):)|file)((?!\$\{).)+?)}'`
to be
`variableSyntax: '\$\{((((self|opt|deep):)|file)((?!\$\{).)+?)}'` (adding `|deep` after `opt`)
it works fine.

Unfortunately, this resulted from a new variable type that I added to deal with recursive deadlock scenarios but was not reflected in your variableSyntax.  Note that your variable syntax also doesn't cover many common new/modern variable types (i.e. s3, ssm, ...).  You might find the `variableSyntax` suggestion at https://serverless.com/framework/docs/providers/aws/guide/variables/#using-custom-variable-syntax useful, although if you are attempting to avoid a syntax change for your serverless variables you may want to simply expand your explicit variable types statement.

I can see that this could be regarded as a bug and breaking change given your configuration.  If you'd like consideration of it as such and would like a change, I'd suggest opening a separate issue and paging me there.

Sorry for the trouble and thank you for facilitating investigation.

**Comment by hojatbay** [comment]
@erikerikson  Thanks for the clarification.

**Comment by erikerikson** [comment]
@bebbi Does https://github.com/serverless/serverless/pull/5156 possibly resolve your issue?  I got a good reproduction of a similar report.  The overwrite syntax was prematurely succeeding with deep values that would later fail to resolve but at that later point the default fallback would no longer be available in the context.

**Comment by pmuens** [comment]
Thanks for opening @bebbi and thanks for providing a fix @erikerikson 👍

I'll close this issue for now since we've pushed more fixes and released a couple of updates. Feel free to re-open if this issue is still around.
